README.md
# RvGeocoder
A Python library for offline reverse geocoding. It improves on an existing library called [reverse_geocoder](https://pypi.org/project/reverse_geocoder/1.5.1/) developed by [Ajay Thampi](https://github.com/thampiman/reverse-geocoder).

The package has not been updated sinse 2016 and this is an effort of keeping it a "live" project.

Main improvements on first version of this forked package:

- support init non-singleton instance of RGeocoder to support different stream simultaneously (creating RGeocoderImpl class to support that)
- added the capability change the header columns when loading custom stream and loaded different schemas
- added query_dist to support returning distance in addtion to the reverse geocoding data itself
- added do_extract to trigger extraction for any Geonames file, e.g: 1000, 15000 etc
- removed win32 and python2 support. added utility function to load data from files/buffers

* The init commit of this project is based on thampiman/reverse-geocoder Latest commit on Sep 15, 2016 a81b4095bf2cb7ef84d2187fcbc8945d5d8922d0 - and forks from there.

## Features
1. Besides city/town and country code, this library also returns the nearest latitude and longitude and also administrative regions 1 and 2.
2. This library also uses a parallelised implementation of K-D trees which promises an improved performance especially for large inputs.

By default, the K-D tree is populated with cities that have a population > 1000. The source of the data is [GeoNames](http://download.geonames.org/export/dump/). You can also load a custom data source so long as it is a comma-separated file with header (like [rg_cities1000.csv](https://github.com/sefiit/rvgeocoder/blob/master/rvgeocoder/rg_cities1000.csv)), containing the following columns:

- `lat`: Latitude
- `lon`: Longitude
- `name`: Name of place
- `admin1`: Admin 1 region
- `admin2`: Admin 2 region
- `cc`: ISO 3166-1 alpha-2 country code

For usage instructions, see below.

## Installation
For first time installation,
```
$ pip install rvgeocoder
```

Or upgrade an existing installation using,
```
$ pip install --upgrade rvgeocoder
```

### Dependencies
1. scipy
2. numpy

### Release Notes
1. v1.0.1 (29-Aug-19) - First version

## Usage
The library supports two modes:

1. Mode 1: Single-threaded K-D Tree (similar to [reverse_geocode](https://pypi.python.org/pypi/reverse_geocode/1.0))
2. Mode 2: Multi-threaded K-D Tree (default)

```python
import rvgeocoder as rvg

coordinates = (41.852968, -87.725730), (48.836364, 2.357422)
print(rvg.search(coordinates))
```

The above code will output the following:
```
[OrderedDict([('lat', '41.84559'),
              ('lon', '-87.75394'),
              ('name', 'Cicero'),
              ('admin1', 'Illinois'),
              ('admin2', 'Cook County'),
              ('cc', 'US')]),
 OrderedDict([('lat', '48.85341'),
              ('lon', '2.3488'),
              ('name', 'Paris'),
              ('admin1', 'Ile-de-France'),
              ('admin2', 'Paris'),
              ('cc', 'FR')])]
```

To use a custom data source for geocoding, you can load the file in-memory and pass it to the library as follows:
```python
import io
import rvgeocoder as rvg

geo = rvg.RGeocoder(stream=io.StringIO(open('custom_source.csv', encoding='utf-8').read()))
coordinates = (51.5214588,-0.1729636),(9.936033, 76.259952),(37.38605,-122.08385)
results = geo.query(coordinates)
```

As mentioned above, the custom data source must be comma-separated with a header as [rg_cities1000.csv](https://github.com/thampiman/reverse-geocoder/blob/master/reverse_geocoder/rg_cities1000.csv).

## Acknowledgements
1. Major inspiration is from Richard Penman's [reverse_geocode](https://bitbucket.org/richardpenman/reverse_geocode) library 
2. First version based on [reverse_geocoder](https://pypi.org/project/reverse_geocoder/1.5.1/) developed by [Ajay Thampi](https://github.com/thampiman/reverse-geocoder)
3. Parallelised implementation of K-D Trees is extended from this [article](http://folk.uio.no/sturlamo/python/multiprocessing-tutorial.pdf) by [Sturla Molden](https://github.com/sturlamolden)
4. Geocoded data is from [GeoNames](http://download.geonames.org/export/dump/)

## License
Copyright (c) 2019 Sefi Itzkovich and contributors. This code is licensed under the LGPL License.


CHANGES.txt
Changes
=======

1.0.7 (2019-09-23)
------------------
- Refactored create_patch_locations and fixed bugs.

1.0.6 (2019-09-18)
------------------
- Refactored code and added RGeocoderDataLoader. Main new feature is create_patch_locations to allow overriding values for specific polygons.

1.0.5 (2019-08-30)
------------------
- support init non-singleton instance of RGeocoder to support different stream simultaneously (creating RGeocoderImpl class to support that)
- added the capability change the header columns when loading custom stream and loaded different schemas
- added query_dist to support returning distance in addtion to the reverse geocoding data itself
- added do_extract to trigger extraction for any Geonames file, e.g: 1000, 15000 etc
- removed win32 and python2 support. added utility function to load data from files/buffers


test.py
from timeit import timeit
import csv
import rvgeocoder as rg

if __name__ == '__main__':
    setup = "import csv;import rvgeocoder as rg;print('Loading coordinates...');" + \
            "cities = [(row[0],row[1]) for row in csv.reader(open('../test/coordinates_10000000.csv','rt'),delimiter='\t')];"
    num = 3
    t = timeit(stmt="rg.search(cities,mode=1)",setup=setup,number=num)
    print('Running time: %.2f secs' % (t / num))

    print('\nLoading coordinates to compare mode 1 and mode 2...')
    cities = [(row[0],row[1]) for row in csv.reader(open('../test/coordinates_1000.csv','rt'),delimiter='\t')]
    result1 = rg.search(cities,mode=1)
    result2 = rg.search(cities,mode=2)
    if result1 != result2:
        print('Results do not match!')
    else:
        print('Both results match!')

tests/test.py
from timeit import timeit
import csv
import rvgeocoder as rvg

if __name__ == '__main__':
    setup = "import csv;import rvgeocoder as rvg;print('Loading coordinates...');" + \
            "cities = [(row[0],row[1]) for row in csv.reader(open('test/coordinates_10000000.csv','rt'),delimiter='\t')];"
    num = 3
    t = timeit(stmt="rvg.search(cities,mode=1)",setup=setup,number=num)
    print('Running time: %.2f secs' % (t / num))

    print('\nLoading coordinates to compare mode 1 and mode 2...')
    cities = [(row[0],row[1]) for row in csv.reader(open('test/coordinates_1000.csv','rt'),delimiter='\t')]
    result1 = rvg.search(cities,mode=1)
    result2 = rvg.search(cities,mode=2)
    if result1 != result2:
        print('Results do not match!')
    else:
        print('Both results match!')

test_loader.py
import csv
import rvgeocoder as rvg
import random
import time

def gen_test_file(n=1000, filename='coordinates_1000.csv'):
    with open('test/%s' % filename, 'w') as f:
        for _ in range(n):
            lat = round(random.uniform(-90,90),2)
            lon = round(random.uniform(-180,180),2)
            f.write('%s\t%s\n' % (lat, lon))

if __name__ == '__main__':
    cities = [(row[0],row[1]) for row in csv.reader(open('test/coordinates_1000000.csv','rt'),delimiter='\t')]
    
    rgeo = rvg.RGeocoderImpl.from_files(['test/file1.csv', 'test/file2.csv'])
    
    start = time.time()
    res = rgeo.query(cities)
    run_time = round(time.time() - start, 2)

    # print first 10 results
    import pprint
    pprint.pprint(res[:10])

    print('** from_files run time is %s' % run_time)


samples/spark_sample.py
import random

import numpy as np
import pandas as pd

import rvgeocoder as rvg
from pyspark.sql.functions import PandasUDFType, col, pandas_udf
from pyspark.sql.session import SparkSession

## TODO: put your custom geocoding files here
files = []

# if we are doing this here mode must be equal to 1 as mode=2 does not work when trying to serialize 
# as ckDTree has some issues with serialization
# files_stream = rvg.RGeocoderDataLoader.load_files_stream(files)
# rgeo = rvg.RGeocoder(mode=1, stream=files_stream)

# return list of strings and not tuple as pandas_udf does not support structs/maps at the moment
def reverse(slat, slon):
    coords = list(zip(slat.values, slon.values))
    if files:
        files_stream = rvg.RGeocoderDataLoader.load_files_stream(files)
        rgeo = rvg.RGeocoder(stream=files_stream)
    else:
        rgeo = rvg.RGeocoder()

    res = rgeo.query(coords)
    
    pdf = pd.DataFrame()
    pdf['locs'] = [[x['cc'], x['name']] for x in res]
    return pdf.locs


def gen_coords_list(n):
    return [(round(random.uniform(-90,90),2), round(random.uniform(-180,180),2)) for i in range(n)]


def main():
    # create spark session
    spark_packages = 'com.amazonaws:aws-java-sdk-pom:1.10.34,org.apache.hadoop:hadoop-aws:2.7.7'
    spark = SparkSession.builder.appName('test_app') \
        .config('spark.jars.packages', spark_packages) \
        .config("spark.hadoop.mapreduce.fileoutputcommitter.algorithm.version", 2) \
        .config("spark.sql.session.timeZone", "UTC") \
        .config("spark.sql.execution.arrow.enabled", "true") \
        .config("spark.sql.execution.arrow.fallback.enabled", "true") \
        .config("spark.sql.execution.arrow.enabled", "true") \
        .getOrCreate()

    # create sample DF, can load from files / streams etc
    sample_size = 10000
    coords = gen_coords_list(sample_size)
    df = spark.createDataFrame(coords, ['latitude', 'longitude'])

    ##################################################################################################
    ##################################################################################################
    ##################################################################################################
    #
    #  THIS IS THE INTERESTING PART - ALL THE REST IS JUST SETUP AND USAGE CODE FOR EXAMPLE
    #
    # run our reverse code that returns a list of country, name (can add also admin1 etc if wanted)
    reverse_udf = pandas_udf(reverse, 'array<string>', PandasUDFType.SCALAR)
    df = df.withColumn('place_names', reverse_udf(df.latitude, df.longitude))
    df = df.select('latitude', 'longitude', col('place_names')[1].alias('name'), col('place_names')[0].alias('cc'))
    #
    #
    ##################################################################################################
    ##################################################################################################
    ##################################################################################################
    
    # Example of usange - print num of coords in US
    places_in_us = df.filter(col('cc') == 'US').count()
    print('Found %s/%s places in US' % (places_in_us, sample_size))


def main_pandas_debug():
    pdf = pd.DataFrame(np.random.randint(0,90,size=(100, 2)), columns=['lat', 'lon'])    
    print(reverse(pdf.lat, pdf.lon))


if __name__ == "__main__":
    main()


rvgeocoder/__init__.py
""" Offline Reverse Geocoder in Python

A Python library for offline reverse geocoding. It improves on an existing library
called reverse_geocode developed by Richard Penman & reverse_geocoder by Ajay Thampi.
"""
from __future__ import print_function

import os
import sys
import csv
csv.field_size_limit(sys.maxsize)
import zipfile
from scipy.spatial import cKDTree as KDTree
from rvgeocoder import cKDTree_MP as KDTree_MP
import numpy as np
import io
from shapely import wkt
from shapely.geometry import Point


GN_URL = 'http://download.geonames.org/export/dump/'
GN_CITIES1000 = 'cities1000'
GN_ADMIN1 = 'admin1CodesASCII.txt'
GN_ADMIN2 = 'admin2Codes.txt'

# Schema of the GeoNames Cities with Population > 1000
GN_COLUMNS = {
    'geoNameId': 0,
    'name': 1,
    'asciiName': 2,
    'alternateNames': 3,
    'latitude': 4,
    'longitude': 5,
    'featureClass': 6,
    'featureCode': 7,
    'countryCode': 8,
    'cc2': 9,
    'admin1Code': 10,
    'admin2Code': 11,
    'admin3Code': 12,
    'admin4Code': 13,
    'population': 14,
    'elevation': 15,
    'dem': 16,
    'timezone': 17,
    'modificationDate': 18
}

# Schema of the GeoNames Admin 1/2 Codes
ADMIN_COLUMNS = {
    'concatCodes': 0,
    'name': 1,
    'asciiName': 2,
    'geoNameId': 3
}

# Schema of the cities file created by this library
RG_COLUMNS = [
    'lat',
    'lon',
    'name',
    'admin1',
    'admin2',
    'cc'
]

# Name of cities file created by this library
RG_FILE = 'rg_cities1000.csv'

# WGS-84 major axis in kms
A = 6378.137

# WGS-84 eccentricity squared
E2 = 0.00669437999014


def singleton(cls):
    """
    Function to get single instance of the RGeocoder class
    """
    instances = {}

    def getinstance(**kwargs):
        """
        Creates a new RGeocoder instance if not created already
        """
        if cls not in instances:
            instances[cls] = cls(**kwargs)
        return instances[cls]
    return getinstance


class RGeocoderDataLoader:
    @classmethod
    def load_files_lines(cls, files: list):
        data_lines = []
        if not files:
            return []
        header_saved = False
        for fl in files:
            with open(fl) as fd:
                # save header only once from the first file
                header = next(fd)
                if not header_saved:
                    data_lines.append(header)
                    header_saved = True
                data_lines.extend(fd.readlines())
        return data_lines

    @classmethod
    def load_files_stream(cls, files: list):
        lines = cls.load_files_lines(files)
        data = ''.join(lines)
        data_stream = io.StringIO(data)
        data_stream.seek(0)
        return data_stream

    @staticmethod
    def _merge_locations(location_files, currect_locations=None):
        common_header = None

        if not currect_locations:
            currect_locations = []
        for loc in location_files:
            with open(loc, 'r') as fd:
                loc_reader = csv.DictReader(fd)
                file_header = loc_reader.fieldnames
                if common_header is None:
                    common_header = file_header
                elif common_header != file_header:
                    raise Exception('File %s has different header than common. Expected header = %s, found = %s' % (
                        loc, common_header, file_header))
                currect_locations.extend(list(loc_reader))
        return currect_locations, common_header

    @staticmethod
    def _remove_polygons_points(locations, polygons_file):
        if not polygons_file:
            return locations

        removed_count = 0
        filtered_locations = []
        polygons = []

        with open(polygons_file, 'r') as fd:
            # get polygons information
            poly_reader = csv.DictReader(fd)
            for row in poly_reader:
                polygons.append((row.get('name', 'unnamed-polygon'),wkt.loads(row['geometry'])))

            # iterate over polygons and remove all points inside of them - avoid collision between sets
            for loc in locations:
                p = Point(float(loc['lon']), float(loc['lat']))
                found = False
                for name, poly in polygons:
                    if poly.contains(p):
                        found = True
                        removed_count += 1
                        print('Removing %s (%s,%s) inside polygon %s' % (
                            loc.get('name', 'unnamed-point'), loc['lat'], loc['lon'], name))
                if not found:
                    filtered_locations.append(loc)
            print('total %s points were removed because found inside patched polygons' % removed_count)
        return filtered_locations

    @classmethod
    def create_patch_locations(cls, location_files: list, patch_loc_file: str,
                               output_file: str = None, patch_poly_file: str = None):
        """ This method recieve a list of location files and two other files describing the patch
        polygon and The points that should represent this polygon in the Spatial index

        REMARK: this can be done much more effiecint and easy using pandas, decided not to use this
        in order to reduce requirements size.
        Arguments:
            location_files {list} -- a csv filename with any schema starting with lat/lon, default is:
                                     lat,lon,name,admin1,admin2,cc
            patch_loc_file {str} -- a csv filename with the same schema as location files
            output_file {str} -- filename for the result of the location files after patching
            patch_poly_file {str} -- OPTIONAL. If given, specifies the a csv file containing polygons of patched
            location and remove all records within these polygon to avoid collision between patch and original 
            location files. schema of file: {cc/name/admin1/admin2}..,geometry(wkt format)
        Returns:
            list of records containing the result of the patched location files
        """

        locations, columns_name = cls._merge_locations(location_files)
        filtered_locations = cls._remove_polygons_points(locations, patch_poly_file)
        new_patched_locations, columns_name = cls._merge_locations([patch_loc_file], filtered_locations)

        if output_file:
            with open(output_file, 'w') as fd:
                writer = csv.DictWriter(fd, fieldnames=columns_name)
                writer.writeheader()
                writer.writerows(new_patched_locations)

        return new_patched_locations


class RGeocoderImpl:
    """
    The main reverse geocoder class
    """
    def __init__(self, mode=2, verbose=True, stream=None, stream_columns=None):
        """ Class Instantiation
        Args:`
        mode (int): Library supports the following two modes:
                    - 1 = Single-threaded K-D Tree
                    - 2 = Multi-threaded K-D Tree (Default)
        verbose (bool): For verbose output, set to True
        stream (io.StringIO): An in-memory stream of a custom data source
        """
        self.mode = mode
        self.verbose = verbose
        if stream:
            coordinates, self.locations = self.load(stream, stream_columns)
        else:
            coordinates, self.locations = self.extract(rel_path(RG_FILE))

        if mode == 1:  # Single-process
            self.tree = KDTree(coordinates)
        else:  # Multi-process
            self.tree = KDTree_MP.cKDTree_MP(coordinates)

    @classmethod
    def from_data(cls, data: str):
        return cls(stream=io.StringIO(data))

    @classmethod
    def from_files(cls, location_files: list):
        """ Loading files data into a stream and creating new instance.
        Arguments:
            location_files {list} -- list of files with lat, lon and additional info on the coord
        Returns:
            [RGeocoderImpl]
        """
        data_stream = RGeocoderDataLoader.load_files_stream(location_files)
        return cls(stream=data_stream)

    def query(self, coordinates):
        """
        Function to query the K-D tree to find the nearest city
        Args:
        coordinates (list): List of tuple coordinates, i.e. [(latitude, longitude)]
        """
        if self.mode == 1:
            _, indices = self.tree.query(coordinates, k=1)
        else:
            _, indices = self.tree.pquery(coordinates, k=1)
        return [self.locations[index] for index in indices]

    def query_dist(self, coordinates):
        """
        Function to query the K-D tree to find the nearest city
        Args:
        coordinates (list): List of tuple coordinates, i.e. [(latitude, longitude)]
        """
        if self.mode == 1:
            dists, indices = self.tree.query(coordinates, k=1)
        else:
            dists, indices = self.tree.pquery(coordinates, k=1)
            # in pquery dists returns a list of arrays so get the first element instead of returning array
            dists = [dist[0] for dist in dists]
        return [(dists[n], self.locations[index]) for (n, index) in enumerate(indices)]

    def load(self, stream, stream_columns):
        """
        Function that loads a custom data source
        Args:
        stream (io.StringIO): An in-memory stream of a custom data source.
                              The format of the stream must be a comma-separated file.
        """
        print('Loading geocoded stream ...')
        stream_reader = csv.DictReader(stream, delimiter=',')
        header = stream_reader.fieldnames

        if stream_columns and header != stream_columns:
            raise csv.Error('Input must be a comma-separated file with header containing ' + \
                'the following columns - %s.\nFound header - %s.\nFor more help, visit: ' % (','.join(stream_columns), ','.join(header)) + \
                'https://github.com/thampiman/reverse-geocoder')

        # Load all the coordinates and locations
        geo_coords, locations = [], []
        for row in stream_reader:
            geo_coords.append((row['lat'], row['lon']))
            locations.append(row)

        return geo_coords, locations

    def extract(self, local_filename):
        """
        Function loads the already extracted GeoNames cities file or downloads and extracts it if
        it doesn't exist locally
        Args:
        local_filename (str): Path to local RG_FILE
        """
        if os.path.exists(local_filename):
            if self.verbose:
                print('Loading formatted geocoded file ...')
            rows = csv.DictReader(open(local_filename, 'rt'))
        else:
            rows = self.do_extract(GN_CITIES1000, local_filename)

        # Load all the coordinates and locations
        geo_coords, locations = [], []
        for row in rows:
            geo_coords.append((row['lat'], row['lon']))
            locations.append(row)
        return geo_coords, locations

    def do_extract(self, geoname_file, local_filename):
        gn_cities_url = GN_URL + geoname_file + '.zip'
        gn_admin1_url = GN_URL + GN_ADMIN1
        gn_admin2_url = GN_URL + GN_ADMIN2

        cities_zipfilename = geoname_file + '.zip'
        cities_filename = geoname_file + '.txt'

        if not os.path.exists(cities_zipfilename):
            if self.verbose:
                print('Downloading files from Geoname...')

            import urllib.request
            urllib.request.urlretrieve(gn_cities_url, cities_zipfilename)
            urllib.request.urlretrieve(gn_admin1_url, GN_ADMIN1)
            urllib.request.urlretrieve(gn_admin2_url, GN_ADMIN2)

        if self.verbose:
            print('Extracting %s...' % geoname_file)
        _z = zipfile.ZipFile(open(cities_zipfilename, 'rb'))
        open(cities_filename, 'wb').write(_z.read(cities_filename))

        if self.verbose:
            print('Loading admin1 codes...')
        admin1_map = {}
        t_rows = csv.reader(open(GN_ADMIN1, 'rt'), delimiter='\t')
        for row in t_rows:
            admin1_map[row[ADMIN_COLUMNS['concatCodes']]] = row[ADMIN_COLUMNS['asciiName']]

        if self.verbose:
            print('Loading admin2 codes...')
        admin2_map = {}
        for row in csv.reader(open(GN_ADMIN2, 'rt'), delimiter='\t'):
            admin2_map[row[ADMIN_COLUMNS['concatCodes']]] = row[ADMIN_COLUMNS['asciiName']]

        if self.verbose:
            print('Creating formatted geocoded file...')
        writer = csv.DictWriter(open(local_filename, 'wt'), fieldnames=RG_COLUMNS)
        rows = []
        for row in csv.reader(open(cities_filename, 'rt'), delimiter='\t', quoting=csv.QUOTE_NONE):
            lat = row[GN_COLUMNS['latitude']]
            lon = row[GN_COLUMNS['longitude']]
            name = row[GN_COLUMNS['asciiName']]
            cc = row[GN_COLUMNS['countryCode']]

            admin1_c = row[GN_COLUMNS['admin1Code']]
            admin2_c = row[GN_COLUMNS['admin2Code']]

            cc_admin1 = cc+'.'+admin1_c
            cc_admin2 = cc+'.'+admin1_c+'.'+admin2_c

            admin1 = ''
            admin2 = ''

            if cc_admin1 in admin1_map:
                admin1 = admin1_map[cc_admin1]
            if cc_admin2 in admin2_map:
                admin2 = admin2_map[cc_admin2]

            write_row = {
                'lat': lat,
                'lon': lon,
                'name': name,
                'admin1': admin1,
                'admin2': admin2,
                'cc': cc}
            rows.append(write_row)
        writer.writeheader()
        writer.writerows(rows)

        if self.verbose:
            print('Removing extracted %s to save space...' % geoname_file)
        os.remove(cities_filename)

        return rows


@singleton
class RGeocoder(RGeocoderImpl):
    pass


def geodetic_in_ecef(geo_coords):
    geo_coords = np.asarray(geo_coords).astype(np.float)
    lat = geo_coords[:, 0]
    lon = geo_coords[:, 1]

    lat_r = np.radians(lat)
    lon_r = np.radians(lon)
    normal = A / (np.sqrt(1 - E2 * (np.sin(lat_r) ** 2)))

    x = normal * np.cos(lat_r) * np.cos(lon_r)
    y = normal * np.cos(lat_r) * np.sin(lon_r)
    z = normal * (1 - E2) * np.sin(lat)

    return np.column_stack([x, y, z])


def rel_path(filename):
    """
    Function that gets relative path to the filename
    """
    return os.path.join(os.getcwd(), os.path.dirname(__file__), filename)


def get(geo_coord, mode=2, verbose=True):
    """
    Function to query for a single coordinate
    """
    if not isinstance(geo_coord, tuple) or not isinstance(geo_coord[0], float):
        raise TypeError('Expecting a tuple')

    _rg = RGeocoder(mode=mode, verbose=verbose)
    return _rg.query([geo_coord])[0]


def search(geo_coords, mode=2, verbose=True):
    """
    Function to query for a list of coordinates
    """
    if not isinstance(geo_coords, tuple) and not isinstance(geo_coords, list):
        raise TypeError('Expecting a tuple or a tuple/list of tuples')
    elif not isinstance(geo_coords[0], tuple):
        geo_coords = [geo_coords]

    _rg = RGeocoder(mode=mode, verbose=verbose)
    return _rg.query(geo_coords)


if __name__ == '__main__':
    print('Testing single coordinate through get...')
    city = (37.78674, -122.39222)
    print('Reverse geocoding 1 city...')
    result = get(city)
    print(result)

    print('Testing coordinates...')
    cities = [(41.852968, -87.725730), (48.836364, 2.357422)]
    print('Reverse geocoding %d locations ...' % len(cities))
    results = search(cities)
    print(results)
